sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/JsonMessageCodec.scala

```diff
@@ -53,20 +57,34 @@ private[kalix] class JsonMessageCodec extends MessageCodec {
     value match {
       case javaPbAny: JavaPbAny   => javaPbAny
       case scalaPbAny: ScalaPbAny => ScalaPbAny.toJavaProto(scalaPbAny)
-      case other                  => JsonSupport.encodeJson(other, lookupTypeHint(other))
+      case other                  => JsonSupport.encodeJson(other, lookupTypeHintWithVersion(other))
     }
   }
 
-  private def lookupTypeHint(value: Any): String =
-    lookupTypeHint(value.getClass)
+  private def lookupTypeHintWithVersion(value: Any): String =
+    lookupTypeHint(value.getClass).currenTypeHintWithVersion
+
+  private[kalix] def lookupTypeHint(clz: Class[_]): TypeHint = {
+    typeHints.computeIfAbsent(clz, computeTypeHint)
+  }
 
-  private[kalix] def lookupTypeHint(clz: Class[_]): String = {
+  private def computeTypeHint(clz: Class[_]): TypeHint = {
     val typeName = Option(clz.getAnnotation(classOf[TypeName]))
       .collect { case ann if ann.value().trim.nonEmpty => ann.value() }
       .getOrElse(clz.getName)
-    cache.computeIfAbsent(clz, _ => typeName)
+
+    val (version, supportedClassNames) = getVersionAndSupportedClassNames(clz)
+    val typeNameWithVersion = typeName + (if (version == 0) "" else "#" + version)
+
     //TODO verify if this could be replaced by sth smarter/safer
-    reversedCache.compute(
+    addToReversedCache(clz, typeName)
```

Comment:
While updating `typeHints` map, I'm also updating another map `reversedTypeHints`, this feels wrong, but I haven't figured out a better option for that.

Reply:
If I recall correctly, we register the types when we register the components. This is supposed to be done by a single thread running over all components and finding out the types. In that sense, it's safe to update the main map and then the reversed one. 

For good measure, we could synchronize this method to make sure we are safe. Just in case something changes one day.